### PR TITLE
Autoloader: Preemptively load unknown plugins from cache

### PR DESCRIPTION
**Pull request by ObliviousHarmony**
Fixes https://github.com/woocommerce/woocommerce/issues/27608

#### Changes proposed in this Pull Request:

One of the limitations of the autoloader is that it is only able to detect plugins using the `active_plugins` option, `active_sitewide_plugins` option, and request parameters. If a plugin is an `mu-plugin` or loaded via code, it will be added to the `$jetpack_autoloader_activating_plugin` global and trigger a reset. This comes with a performance penalty as well as the potential for odd behavior when two classes from different versions of the same package are autoloaded.

This PR introduces a cache file in the `wp-content/cache` folder containing a list of plugins that are known to the autoloader on shutdown. By saving this cache on shutdown we're able to preemptively handle autoloaders in plugins that we aren't aware of initially. 

This solution is not perfect however, as it causes the autoloader to consider plugins that may not necessarily be active anymore. We aim for eventual consistency instead by only writing plugins that we've discovered to the cache, so the request after a plugin is deactivated, it will be removed from the cache. My thinking is that the presence of a deactivated plugin's class files shouldn't cause any problems. One odd thing that might happen is that a newer autoloader is used in a deactivated plugin, however, the cache invalidation will mean this only happens a single time.

I've also taken the opportunity to finish the refactor I started in #16819. One of the core motivations behind this refactoring was to decrease the cognitive cost of working with the autoloader. Although the code may be simple it hid a lot of complexity and required deep thought into how different elements might interact. Here are some highlights:
* High-level integration tests that check common cross-plugin conflict scenarios
* `CacheHandler` for reading/writing filesystem cache files. This can enable other optimizations like saving the reconciled classmap.
* `Container` for managing all of the dependencies in the package.
* `Plugin_Locator` to centralize all of the plugin lookup logic and generalize certain shared actions
* `Autoloader` to serve as a new entry point
* `Hook_Manager` shared by all autoloaders so that resets can remove hooks added by other autoloaders

#### Jetpack product discussion
N/A

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:

* Make sure Jetpack works by itself
* Make sure Jetpack works alongside WooCommerce
* Make sure Jetpack as an `mu-plugin` works alongside WooCommerce

There's arguably a large matrix of possible ways to test the autoloader, which I attempted to cover most if not all of in the `Test_Autoloader_Scenarios` integration tests. If you see any tests that we should add, please note!

#### Proposed changelog entry for your changes:

* Autoloader: Added support for preemptively loading unknown plugins from a cache file.

### CODE REVIEW

File: packages/autoloader/src/class-autoloader.php
Comment (kbrown9) [inline packages/autoloader/src/class-autoloader.php:63]: Should the cache be cleared if an error happens during a plugin's initialization?

For example, a plugin could load a version of a package that conflicts with another plugin, and the problem plugin could be deactivated. However, it would still be in the cache and its packages could still be loaded, right?

File: packages/autoloader/src/class-container.php
Comment (kbrown9) [inline packages/autoloader/src/class-container.php:65]: I thought about the shared `Hook_Manager` object a lot. I'm not enthusiastic about the fact that we could use an older version of the `Hook_Manager` class if an older version of the autoloader is registered first. However, I think that would just be a temporary situation, and it would be resolved on the next page load, right?

This situation would only happen when the plugin with the latest autoloader was not in the active plugins list (for example, it's an mu-plugin or it's activating in a way that couldn't be detected). The plugin would be added to the cache file. Then its autoloader would be identified as the latest autoloader in subsequent page loads, and the latest version of the `Hook_Manager` class would be used. Is that correct?

File: packages/autoloader/src/class-plugins-handler.php
Comment (kbrown9) [inline packages/autoloader/src/class-plugins-handler.php:72]: What does "activated via the interface" mean? I think the `activate_plugins` option contains all of the currently active plugins, regardless of how they were activated.

File: packages/autoloader/src/class-plugin-locator.php
Comment (kbrown9) [inline packages/autoloader/src/class-plugin-locator.php:17]: `get_site_option` calls `get_network_option`, which calls `get_option` when the site is not a multisite (see [here](https://core.trac.wordpress.org/browser/tags/5.5.1/src/wp-includes/option.php#L1380)). Do you think we should consider removing the `$site_option` parameter and just using `get_site_option`?

**Review by kbrown9 — COMMENTED** [review]
I completed a first pass of the code changes, and I just have a couple of questions. I haven't manually tested or thoroughly reviewed the unit/integration tests yet.

File: packages/autoloader/src/class-autoloader-handler.php
Comment (ObliviousHarmony) [reply packages/autoloader/src/class-autoloader-handler.php:72]: Good catch!

File: packages/autoloader/src/class-hook-manager.php
Comment (ObliviousHarmony) [reply packages/autoloader/src/class-hook-manager.php:48]: Since the `Hook_Manager` is intended to be a shared global I thought it would be wise to add the rest of its API to avoid potential bugs if filter support was added. To be honest I don't see a use-case for filters in the autoloader given its utility but I didn't want to rule out the possibility since it's a small function.

File: packages/autoloader/src/class-autoloader.php
Comment (ObliviousHarmony) [reply packages/autoloader/src/class-autoloader.php:63]: That seems like a good precaution to take. The goal here was to avoid cases where the cache would lead to further errors and that adds to it.

File: packages/autoloader/src/class-plugin-locator.php
Comment (ObliviousHarmony) [reply packages/autoloader/src/class-plugin-locator.php:17]: My preference would be that we remain explicit here since `get_site_option` also calls some filters and can be short-circuited with the expectation that it's a multisite installation. It feels like using `get_site_option` would be masking these details that might be relevant.

File: packages/autoloader/src/class-plugins-handler.php
Comment (ObliviousHarmony) [reply packages/autoloader/src/class-plugins-handler.php]: Who knows :) I've adjusted the comment!

File: packages/autoloader/src/class-container.php
Comment (ObliviousHarmony) [reply packages/autoloader/src/class-container.php:65]: > I'm not enthusiastic about the fact that we could use an older version of the Hook_Manager class if an older version of the autoloader is registered first.

My rationale here was that it seemed incredibly unlikely we would ever change the `Hook_Manager` class. I'd considered some form of version reconciliation but given how "complete" the autoloader feels I figured we should cross that bridge if we get there.

> Is that correct?

Yep! The only caveat is that the newer autoloader will need to be aware of that limitation and act accordingly.

**Proposed Solution**

The solution to this problem I'd entertained was an interface for all shared dependencies that declared a `getProperties()` method and a `VERSION` constant string defined as part of the autoloader generation. The idea is that the constructor of newer classes can consume the older class and receive its properties in order to replace it.

If you think this is something we should do now though I'm not opposed to adding it.

**Comment by ObliviousHarmony** [comment]
I've addressed all of your feedback, thank you very much for the initial review @kbrown9!

**Comment by ObliviousHarmony** [comment]
After chatting with @WPprodigy, I've opted to replace the `Cache_Handler` with the Transient API. This will perform better in practice because the transients will be autoloaded, as well as better support read-only filesystems. I plan to make another change tomorrow to tokenize the paths since the cache is shared across webservers, but I've added a commit with the initial change for now!

**Comment by ObliviousHarmony** [comment]
I've finished making the adjustments to use the Transient API with tokenized paths in the cache. Since this PR includes some enhancements to the test suite I'd like to wait for #17813 to land so that I can adjust the scenario tests to generate the autoloader correctly.

**Comment by kbrown9** [comment]
> I've finished making the adjustments to use the Transient API with tokenized paths in the cache. Since this PR includes some enhancements to the test suite I'd like to wait for #17813 to land so that I can adjust the scenario tests to generate the autoloader correctly.

#17813 has landed.

**Comment by ObliviousHarmony** [comment]
Thanks @kbrown9!

I'm actually just going to keep this as-is for now in favor of perhaps looking at removing the Composer autoload generation and enhancing some of the "plugin" aspects of our Composer plugin.

**Comment by WPprodigy** [comment]
My main concern is around the edge conditions, so:

- A plugin is added to the site (one of the non traditional ways) and not recognized on the initial load and causes a fatal as a result. Does it still get added to this cache so the next request is able to load it in successfully?
- That plugin is then removed from the site, is the cache going to gracefully handle it no longer existing?

**Comment by ObliviousHarmony** [comment]
We can add tests for this too I imagine but regarding what it SHOULD do:

> A plugin is added to the site (one of the non traditional ways) and not recognized on the initial load and causes a fatal as a result. Does it still get added to this cache so the next request is able to load it in successfully?

If the fatal was thrown before `plugins_loaded` then it will invalidate the cache entirely. If a fatal is thrown after, then the plugin will be added. Our goal is only to ensure that fatals thrown due to bad autoloader behavior don't result in looping fatals.

Regarding the error in Woo, we are now caching the active list in a global and perform a reset if that list changes due to option filtering.

> That plugin is then removed from the site, is the cache going to gracefully handle it no longer existing?

If it is removed then the manifests won't be found and it will remove the plugin from the cache on shutdown that request. If it is deactivated then it will load the class files into its manifest the first request and remove it from the cache. This should be harmless since it isn't actually doing anything and will be consistent after.

**Comment by ObliviousHarmony** [comment]
I've added some test coverage around the cache update behavior @WPprodigy. This should be a good enough approximation of what takes place during an early shutdown, such as in the case of a fatal during initialization.

@kbrown9 I think this is ready for any final review now!

File: packages/autoloader/src/class-plugins-handler.php
Comment (kbrown9) [inline packages/autoloader/src/class-plugins-handler.php:45]: I think we need to normalize the current plugin's path with `str_replace( '\\', '/', dirname( dirname( __DIR__ ) );` here.

File: packages/autoloader/src/class-plugin-locator.php
Comment (kbrown9) [inline packages/autoloader/src/class-plugin-locator.php:102]: I think we need to normalize the plugin path with `str_replace( '\\', '/', $path )` here.

File: packages/autoloader/src/class-plugin-locator.php
Comment (kbrown9) [inline packages/autoloader/src/class-plugin-locator.php:107]: `str_replace( '\\', '/', $path )` should also be called here.

**Review by kbrown9 — CHANGES_REQUESTED** [review]
I did some testing on a Windows server, and I noticed that the Jetpack plugin path was added to the cached list of active plugins twice. It looks like we need to normalize the plugin paths to ensure that the path separators are consistent on Windows servers. See my in-line comments.

**Comment by kbrown9** [comment]
I've done some additional manual testing, and everything seemed to work well. I tested the following scenarios:

 * Jetpack alone installed as a normal plugin.
 * Jetpack, WooCommerce, and WooCommerce Payments installed as normal plugins.
 * Jetpack installed as an MU plugin, Woo and Woo Payments installed as normal plugins.
 * Jetpack and Woo installed as MU plugins, Woo Payments installed as a normal plugin.
 * Jetpack, Woo, and Woo Payments all using the in-development autoloader installed as normal plugins.

File: packages/autoloader/src/class-hook-manager.php
Comment (kbrown9) [reply packages/autoloader/src/class-hook-manager.php:48]: We would normally remove unused code, but I think it's ok to include it in this case.

File: packages/autoloader/src/class-container.php
Comment (kbrown9) [reply packages/autoloader/src/class-container.php:65]: I don't think it's necessary to do that now. As you pointed out, it seems unlikely that `Hook_Manager` will need to be changed.

File: packages/autoloader/src/class-plugin-locator.php
Comment (ObliviousHarmony) [reply packages/autoloader/src/class-plugin-locator.php:102]: The path that comes out of the processor is already normalized, so this likely isn't the cause.

Comment (ObliviousHarmony) [reply packages/autoloader/src/class-plugin-locator.php:102]: Sorry, more accurately that it's supposed to be, and so a lack of normalized paths elsewhere means it isn't handled properly there!

File: packages/autoloader/src/class-plugins-handler.php
Comment (ObliviousHarmony) [reply packages/autoloader/src/class-plugins-handler.php]: I've gone ahead and moved this functionality into the `Plugin_Locator` since it seems to fit some of the dependencies already there as well as the general theme of the class. It's now normalized correctly as a result!

**Comment by ObliviousHarmony** [comment]
Thank you @kbrown9 for the thorough review as well as the confidence check with the plugin testing. I've added path normalization to the `Path_Processor` and some unit tests for Windows paths in a few key places. This should be sufficient as all of the plugin paths pass through the processor.

You can check out those changes in https://github.com/Automattic/jetpack/pull/17667/commits/9c201d119f96a85a51ce2ea4536eb8ec47a2e67f. 

Let me know if you're still seeing the double cache entries on Windows! You may need to clear the `jetpack_autoloader_plugin_paths` transient, as I'm not entirely sure whether or not it will update the invalid cache in this case.

**Comment by kraftbj** [comment]
I brought in the latest from `master` to pick up the PHP 8 compat commit just merged so tests will pass.

**Review by kbrown9 — APPROVED** [review]
This tests well for me! I'm no longer seeing problems on Windows.

**Comment by ObliviousHarmony** [comment]
@kbrown9 I've merged in `master` to grab the latest set of PHP 8 compatibility fixes and made the changes in the tests accordingly! The merge conflicts have all been resolved and this should be done!

**Comment by kbrown9** [comment]
> @kbrown9 I've merged in `master` to grab the latest set of PHP 8 compatibility fixes and made the changes in the tests accordingly! The merge conflicts have all been resolved and this should be done!

It looks like there are a few PHP 8 warnings in this test: https://github.com/Automattic/jetpack/pull/17667/checks?check_run_id=1512979024

**Review by kbrown9 — APPROVED** [review]

**Comment by kbrown9** [comment]
I merged with the latest master to bring in some additional changes to the required checks. I'll re-approve when the tests are complete, and then we can merge this!

**Review by kbrown9 — APPROVED** [review]

**Pull request merged**
